core/Pos/Binary/Core/Types.hs

```diff
@@ -5,54 +5,53 @@ import           Universum
-                                             UnsignedVarInt (..), deriveSimpleBi, label,
-                                             labelP, labelS, putField, putWord8)
-import qualified Pos.Binary.Core.Coin       as BinCoin
+import           Pos.Binary.Class           (Bi (..), Cons (..), Field (..), deriveSimpleBi)
 import           Pos.Binary.Core.Fee        ()
 import           Pos.Binary.Core.Script     ()
 import           Pos.Binary.Core.Version    ()
+import qualified Pos.Binary.Core.Coin       as BinCoin
 import qualified Pos.Core.Fee               as T
 import qualified Pos.Core.Types             as T
 import qualified Pos.Data.Attributes        as A
-import           Pos.Util.Util              (eitherToFail)
 
 -- kind of boilerplate, but anyway that's what it was made for --
 -- verbosity and clarity
 
 instance Bi T.Timestamp where
-    sizeNPut = labelS "Timestamp" $ putField toInteger
-    get = label "Timestamp" $ fromInteger <$> get
+  encode (T.Timestamp ms) = encode . toInteger $ ms
+  decode = T.Timestamp . fromIntegral <$> decode @Integer
 
 instance Bi T.TimeDiff where
-    sizeNPut = labelS "TimeDiff" $ putField toInteger
-    get = label "TimeDiff" $ fromInteger <$> get
+    encode = encode . toInteger
+    decode = fromInteger <$> decode
 
 instance Bi T.EpochIndex where
-    sizeNPut = labelS "EpochIndex" $ putField (UnsignedVarInt . T.getEpochIndex)
-    get = label "EpochIndex" $ T.EpochIndex . getUnsignedVarInt <$> get
+  encode (T.EpochIndex epoch) = encode epoch
+  decode = T.EpochIndex <$> decode
 
 instance Bi (A.Attributes ()) where
-    size = VarSize $ A.sizeAttributes (\() -> [])
-    get = label "Attributes" $
-        A.getAttributes (\_ () -> Nothing) (Just (128 * 1024 * 1024)) ()
-    put = labelP "Attributes" . A.putAttributes (\() -> [])
+  encode = A.encodeAttributes []
+  decode = A.decodeAttributes () $ \_ _ _ -> Nothing
 
 instance Bi T.Coin where
-    size = VarSize BinCoin.size
-    put = labelP "Coin" . mapM_ putWord8 . BinCoin.encode
-    get = label "Coin" $ BinCoin.decode
+  encode = encode . BinCoin.encode
+  decode = BinCoin.decode
 
 instance Bi T.CoinPortion where
-    sizeNPut = labelS "CoinPortion" $ putField T.getCoinPortion
-    get = label "CoinPortion" $ get >>= T.mkCoinPortion
+  encode = encode . T.getCoinPortion
+  decode = do
+    word64 <- decode @Word64
+    case T.mkCoinPortion word64 of
+      Left err          -> fail err
+      Right coinPortion -> return coinPortion
 
 instance Bi T.LocalSlotIndex where
-    sizeNPut = labelS "LocalSlotIndex" $
-        putField (UnsignedVarInt . T.getSlotIndex)
-    get =
-        label "LocalSlotIndex" $
-        eitherToFail . T.mkLocalSlotIndex . getUnsignedVarInt =<< get
+  encode = encode . T.getSlotIndex
+  decode = do
+    word16 <- decode @Word16
+    case T.mkLocalSlotIndex word16 of
+      Left err        -> fail ("decode@LocalSlotIndex: " <> toString err)
```

Comment:
I have no strong opinions, if not that maybe having an extra error prefix might help pinpointing things in the code. What do you reckon?